rfcs/text/0033/process.yml

```diff
@@ -0,0 +1,61 @@
+# Licensed to Elasticsearch B.V. under one or more contributor
+# license agreements. See the NOTICE file distributed with
+# this work for additional information regarding copyright
+# ownership. Elasticsearch B.V. licenses this file to you under
+# the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# 	http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+---
+- name: process
+  fields:
+
+    # minor typo fix in description for this field. "pts/0 -> "pts/0"
+    - name: tty.char_device.major
+      description: >
+        The major number identifies the driver associated with the device. The character device's major and minor numbers can be algorithmically combined to produce the more familiar terminal identifiers such as "ttyS0" and "pts/0". For more details, please refer to the Linux kernel documentation.
+
+    # new 'io' properties for tracking tty output
+    - name: io
```

Comment:
@mitodrummer given that we want this to be extensible for different io sources, and my assumption is that the event type will be "io", does it make more sense for this to be an array of objects? so that different io.sources can piggyback on the same process event.

Reply:
Might be cleaner to keep different sources in separate events? Perhaps we can use event.action as the source differentiator instead of process.io.source ? e.g event.action: 'tty_io|file_io|network_io'.

Reply:
That'll work too, but then 'io.source' becomes redundant? I'm also ok with this as is, it looks good, just wanted to bring this up wrt searchability and UI.

Reply:
yea, if we use event.action to differentiate "sources" we can drop process.io.source.

Reply:
Though not applicable to ECS, there is also the question of which "index" we want to write to? At the moment I'm assuming it will all go into logs-endpoint.events.process. Whether that is a good idea or not needs answering (perhaps from people on the endpoint side). Do we want to consider a new data stream? e.g  logs-endpoint.events.tty ?

Reply:
if we generalized io to be for any fd {file/socket/..} and for both reads and writes (input & output - IO), we'd still want it to align with the process.* fields that indicate which process did the writing and reading to those/that fd.   We'd likely need to keep timestamps for reads and writes to make sense of the stream too.   We'd probably have fine-grained controls on which programs or file/sockets/...  we actually track read and maybe writes for for data volume mgmt reasons.

Ideally this stream of events is serialized or has timing information for reads and writes (e.g. I could see that this read from a TCP socket was after this write of the request)

Off the top of my head, the simplest way to render these to events would be have each IO event represent  (Reads xor Writes) (never both), on a single tty/file/socket/... by a single process.   This approach would allow the event.action to be "input" or "output" and also only need one source per event (only be the controlling tty supported for now but eventually files, sockets, etc could fit into this model).   If we flush these events (ie stop aggregating read xor written data to a specific tty/file/socket) when the operation changes from read to write or write to read on the same tty/file/socket we could get an event storm from some types of program but likely rare?)

Reply:
"Ideally this stream of events is serialized or has timing information for reads and writes". Each entry in the data array could include this timing info (like suggested to support quasi realtime playback). Could the write mode be set at that level too?
e.g
process.io.data = [
  { time: <timestamp>, mode: 'w', raw: 'the raw data written' },
  { time: <timestamp>, mode: 'r', raw: 'the raw data read' }
];

this would solve the issue of frequent operation changes perhaps. Of course we run into the limitation elastic has on arrays of objects, so would be harder to query for just input or output events.

Reply:
I forgot to mention the search-ability should be precise (for rules generating alerts via queries on ES and threat hunting) e.g. - matching "output to this tty/file/socket containing this text" which is problem with the r/w mode in the array :(. So a separate event (r xor w) seems easiest - or a separate array. 

General purpose io support would also require b64 or something similar for binary data that cant be converted to utf8 or scoping these io events to text only capture (ie never send up binary io).

Reply:
to support binary data maybe we have ability to set either a "text" property or "binary/bytes" field.
e.g
For tty io
```
process.io.data = [
  { text: 'utf8 encoded text' },
  { text: 'utf8 encoded text' }
];
```
For binary data
```
process.io.data = [
  { bytes: <base64encodedvalue> },
  { bytes: <base64encodedvalue> }
];
```
Field types would be:
process.io.data.text = keyword|wildcard type
process.io.data.bytes = binary type (optimized type for storing base64 encoded bytes) <- i imagine this type skips any sort of tokenization or indexing for search.

**event.action** could then inform applications which data their dealing with. e.g tty_input, tty_output, file_input, file_output etc..

Reply:
The text entries in the array represent lines, ideally.  No such constraint possible for binary so likely don't need an array for it?  Also for event action, I'm not sure the source/dest type needs to be there? e.g event.action = {binary_input, binary_output, text_input, text_output} and the src/dst would be detailed in the event in other fields.

so 

```
event.action = {binary_input, binary_output, text_input, text_output} 

process.io.bytes = one base64 string the of read/input written/output binary data. Popluated for binary_input and binary_output event.action values.

process.io.data.[].txt = each array entry is a string, with best efforts to line-ify the data, including a trailing newline to indicate a complete line. Populated for text_input and text_output event.action values.